### PR TITLE
driver/usbvideodriver: add basic MacroSilicon USB Video support

### PR DESCRIPTION
**Pull request by jluebbe**
This device only support JPEG.

**Checklist**
- [ ] Documentation for the feature
- [ ] Tests for the feature 
- [x] PR has been tested
- [ ] Man pages have been regenerated

### CODE REVIEW

File: labgrid/driver/usbvideodriver.py [hunk 1]
```diff
@@ -54,7 +62,8 @@ def stream(self, caps_hint=None):
         tx_cmd = self.video.command_prefix + ["gst-launch-1.0", "-q"]
         tx_cmd += pipeline.format(self.video.path, caps).split()
         rx_cmd = ["gst-launch-1.0"]
-        rx_cmd += "fdsrc ! h264parse ! avdec_h264 ! glimagesink sync=false".split()
+        #rx_cmd += "fdsrc ! h264parse ! avdec_h264 ! glimagesink sync=false".split()
+        rx_cmd += "fdsrc ! jpegparse ! jpegdec ! glimagesink sync=false".split()
```
Comment (jluebbe) [inline labgrid/driver/usbvideodriver.py:66]: This needs to be more flexible. Maybe return more information from _get_caps (encoder+decoder)?

Comment (Emantor) [reply labgrid/driver/usbvideodriver.py:66]: Yes, sgtm.

Comment (jluebbe) [reply labgrid/driver/usbvideodriver.py]: I've talked with @michaelolbrich about this, the simpler approach is to use mkv and let the client use playbin.

File: labgrid/driver/usbvideodriver.py [hunk 2]
```diff
@@ -54,7 +63,7 @@ def stream(self, caps_hint=None):
         tx_cmd = self.video.command_prefix + ["gst-launch-1.0", "-q"]
         tx_cmd += pipeline.format(self.video.path, caps).split()
         rx_cmd = ["gst-launch-1.0"]
-        rx_cmd += "fdsrc ! h264parse ! avdec_h264 ! glimagesink sync=false".split()
+        rx_cmd += "playbin uri=fd://0 sync=false".split()
```
Comment (michaelolbrich) [inline labgrid/driver/usbvideodriver.py:66]: `playbin` has no `sync` property. And it should not be necessary here anyways.

Comment (jluebbe) [reply labgrid/driver/usbvideodriver.py]: OK. I've removed the sync=false.

**Comment by jluebbe** [comment]
@Emantor @michaelolbrich I've also added support for passing v4l controls.

File: labgrid/driver/usbvideodriver.py [hunk 3]
```diff
@@ -38,23 +44,36 @@ def select_caps(self, hint=None):
...
-            return "v4l2src device={} ! {} ! decodebin ! vaapipostproc ! vaapih264enc ! h264parse ! fdsink"  # pylint: disable=line-too-long
-        raise InvalidConfigError("Unkown USB video device {:04x}:{:04x}".format(*match))
+            controls = controls or "focus_auto=1"
+            inner = "h264parse"
+        elif match == (0x046d, 0x0892):
+            inner = "decodebin ! vaapipostproc ! vaapih264enc ! h264parse"
+        elif match == (0x534d, 0x2109):
+            inner = None  # just forward the jpeg frames
+        else:
+            raise InvalidConfigError("Unkown USB video device {:04x}:{:04x}".format(*match))
```
Comment (michaelolbrich) [inline labgrid/driver/usbvideodriver.py:57]: You could allow arbitrary cameras with something like this:
`v4l2src device=/dev/video0 ! 'image/jpeg;video/x-h264' ! parsebin ! matroskamux ! streamable=true ! fdsink`

Note that the order of the caps define the priority. So in this example, jpeg will be used if the camera supports both jpeg and h264.

Comment (michaelolbrich) [reply labgrid/driver/usbvideodriver.py:57]: And you need some caps here, otherwise some raw format will most likely be chosen.

Comment (jluebbe) [reply labgrid/driver/usbvideodriver.py:57]: We currently select the caps explicitly from low/mid/high settings.

Comment (michaelolbrich) [reply labgrid/driver/usbvideodriver.py:57]: You can use ranges there: `image/jpeg,width=[1,1280]height=[1,720],framerate=[1/1,15/1]`
From what I can tell, it tries to use the larges values that fit into the range.

Comment (jluebbe) [reply labgrid/driver/usbvideodriver.py:57]: OK. But for now, I'd prefer to use tested and explicit values there.

**Review by Emantor — DISMISSED** [review]

**Comment by Emantor** [comment]
LGTM, have not tested yet.

File: labgrid/driver/usbvideodriver.py [hunk 4]
```diff
@@ -38,23 +44,36 @@ def select_caps(self, hint=None):
...
         match = (self.video.vendor_id, self.video.model_id)
         if match == (0x046d, 0x082d):
-            return "v4l2src device={} ! {} ! h264parse ! fdsink"
-        if match == (0x046d, 0x0892):
-            return "v4l2src device={} ! {} ! decodebin ! vaapipostproc ! vaapih264enc ! h264parse ! fdsink"  # pylint: disable=line-too-long
-        raise InvalidConfigError("Unkown USB video device {:04x}:{:04x}".format(*match))
+            controls = controls or "focus_auto=1"
+            inner = "h264parse"
+        elif match == (0x046d, 0x0892):
+            inner = "decodebin ! vaapipostproc ! vaapih264enc ! h264parse"
```
Comment (Emantor) [inline labgrid/driver/usbvideodriver.py:53]: This breaks this pipeline, however even on master this is currently displaying in green, do i'd be in favor of moving this to jpeg too and reworking the vaapi support eventually.

**Review by Emantor — APPROVED** [review]

File: labgrid/driver/usbvideodriver.py (hunk 4 above)
Comment (jluebbe) [reply labgrid/driver/usbvideodriver.py]: @Emantor has switched this to jpeg.

**Pull request merged**